hoomd/BoxResizeUpdater.cc

```diff
@@ -45,105 +39,142 @@ BoxResizeUpdater::~BoxResizeUpdater()
+    auto final_box = getBoxDimFromPyObject(m_final_box);
+    Scalar3 L1 = initial_box.getL();
+    Scalar3 L2 = final_box.getL();
+    Scalar Lx = L2.x * scale + (1.0 - scale) * L1.x;
+    Scalar Ly = L2.y * scale + (1.0 - scale) * L1.y;
+    Scalar Lz = L2.z * scale + (1.0 - scale) * L1.z;
+    Scalar xy = final_box.getTiltFactorXY() * scale +
+                (1.0 - scale) * initial_box.getTiltFactorXY();
+    Scalar xz = final_box.getTiltFactorXZ() * scale +
+                (1.0 - scale) * initial_box.getTiltFactorXZ();
+    Scalar yz = final_box.getTiltFactorYZ() * scale +
+                (1.0 - scale) * initial_box.getTiltFactorYZ();
+
+    BoxDim new_box = BoxDim(make_scalar3(Lx, Ly, Lz));
+    new_box.setTiltFactors(xy, xz, yz);
+    return new_box;
     }
 
-/*! Perform the needed calculations to scale the box size
+/** Perform the needed calculations to scale the box size
     \param timestep Current time step of the simulation
 */
 void BoxResizeUpdater::update(unsigned int timestep)
     {
     m_exec_conf->msg->notice(10) << "Box resize update" << endl;
     if (m_prof) m_prof->push("BoxResize");
 
-    // first, compute what the current box size and tilt factors should be
-    Scalar Lx = (*m_Lx)(timestep);
-    Scalar Ly = (*m_Ly)(timestep);
-    Scalar Lz = (*m_Lz)(timestep);
-    Scalar xy = (*m_xy)(timestep);
-    Scalar xz = (*m_xz)(timestep);
-    Scalar yz = (*m_yz)(timestep);
+    // first, compute the new box
+    BoxDim new_box = getCurrentBox(timestep);
 
     // check if the current box size is the same
-    BoxDim curBox = m_pdata->getGlobalBox();
-    Scalar3 curL = curBox.getL();
-    Scalar curxy = curBox.getTiltFactorXY();
-    Scalar curxz = curBox.getTiltFactorXZ();
-    Scalar curyz = curBox.getTiltFactorYZ();
-
-    // copy and setL + setTiltFactors instead of creating a new box
-    BoxDim newBox = curBox;
-    newBox.setL(make_scalar3(Lx, Ly, Lz));
-    newBox.setTiltFactors(xy,xz,yz);
-
-    bool no_change = fabs((Lx - curL.x) / Lx) < 1e-7 &&
-                     fabs((Ly - curL.y) / Ly) < 1e-7 &&
-                     fabs((Lz - curL.z) / Lz) < 1e-7 &&
-                     fabs((xy - curxy) / xy) < 1e-7 &&
-                     fabs((xz - curxz) / xz) < 1e-7 &&
-                     fabs((yz - curyz) / yz) < 1e-7;
+    BoxDim cur_box = m_pdata->getGlobalBox();
 
     // only change the box if there is a change in the box dimensions
-    if (!no_change)
+    if (new_box != cur_box)
         {
         // set the new box
-        m_pdata->setGlobalBox(newBox);
+        m_pdata->setGlobalBox(new_box);
 
         // scale the particle positions (if we have been asked to)
         if (m_scale_particles)
             {
             // move the particles to be inside the new box
-            ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(), access_location::host, access_mode::readwrite);
+            ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(),
```

Comment:
The `BoxResizer` runs on a trigger like other `Updaters`, so it does not run every step. And yes, currently if running on the GPU we must copy the memory and use the CPU for every timestep it is triggered, if we resize. There are plans in the future to create a GPU kernel to do this to prevent this copy and improve performance, but I decided to just port the existing functionality for now to get the beta out there earlier.

I want to implement the GPU kernel later to get some GPU coding experience, but will probably wait until later in HOOMD v3's development.